src/main/java/org/opensearch/knn/index/query/rescore/RescoreSearcher.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query.rescore;
+
+import lombok.AllArgsConstructor;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.Bits;
+import org.opensearch.knn.index.query.KNNWeight;
+import org.opensearch.knn.index.query.ResultUtil;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.stream.Collectors;
+
+@AllArgsConstructor
+public final class RescoreSearcher {
+
+    private final ResultUtil resultUtil;
+
+    /**
+     * Executes a k-NN query by executing a first pass search over an index with quantized vectors with
+     * an oversample factor and then rescoring down to k results for the whole shard.
+     *
+     * @param indexSearcher The index searcher for the current search
+     * @param leafReaderContexts The leaf reader contexts for the current search
+     * @param knnWeight The knn weight for the current search
+     * @return The top k results for the shard
+     * @throws IOException If an error occurs during the search
+     */
+    public TopDocs[] search(
+        final IndexSearcher indexSearcher,
+        List<LeafReaderContext> leafReaderContexts,
+        KNNWeight knnWeight,
+        int firstPassK,
+        int finalK
+    ) throws IOException {
+        List<Map<Integer, Float>> perLeafResults = firstPassSearch(indexSearcher, leafReaderContexts, knnWeight, firstPassK);
+        return rescoreResults(indexSearcher, leafReaderContexts, knnWeight, perLeafResults, finalK);
```

Comment:
should we add some debug logs which logs the time taken to run these 2 functions. I know there is no good way to record time for these queries. But if we can add some debug logs they might come handy later on.

Reply:
Debug logs are good. Which method did you wrap with a timer before? I can model it based off of that

Reply:
I think I wrap functions in a feature branch and not in the main branch. So the code might not be a good. :P 

I had this code: https://github.com/navneet1v/k-NN/blob/perf-deep-dive/src/main/java/org/opensearch/knn/jni/JNIService.java#L110-L117 long time back. But not a good way I think.

Reply:
I thought you added an annotation somewhere. Might be mistaken.

Reply:
I created the GH issue for the work. But I didn't pick that up. 

Ref: https://github.com/opensearch-project/k-NN/issues/1594